File: src/main/csharp/FalsePositive/Loop/LoopCondition_FP_ImmutableConstants.cs
using System;
using System.Web;
using System.Web.UI;

namespace Checkmarx.FalsePositive.Loop
{
    /// <summary>
    /// FALSE POSITIVE SCENARIOS: Unchecked Input for Loop Condition - Immutable Constants (C#)
    /// 
    /// Pattern: Loop bounds are constrained by readonly constants, ensuring safe iteration counts
    /// 
    /// Even though user input flows to the loop condition, it's bounded by immutable constants.
    /// CxQL doesn't recognize Math.Min/Max with constants as safe bounds.
    /// 
    /// All scenarios are FALSE POSITIVES - SAFE bounded loops.
    /// </summary>
    public class LoopCondition_FP_ImmutableConstants : Page
    {
        private const int MaxItems = 100;
        private const int MaxPageSize = 50;
        private const int MaxRetries = 10;
        private const int MaxBatchSize = 1000;
        private const long MaxIterations = 500L;
        
        /// <summary>
        /// FALSE POSITIVE: Math.Min() with const
        /// </summary>
        protected void ProcessMathMinBounded()
        {
            string countParam = Request.QueryString["count"];
            int userCount = int.Parse(countParam);
            
            int safeCount = Math.Min(userCount, MaxItems);
            
            for (int i = 0; i < safeCount; i++) // FALSE POSITIVE - Bounded
            {
                Response.Write("Item " + i + "<br/>");
            }
        }
        
        /// <summary>
        /// FALSE POSITIVE: Ternary with const
        /// </summary>
        protected void ProcessTernaryBounded()
        {
            string sizeParam = Request.QueryString["pageSize"];
            int userSize = int.Parse(sizeParam);
            
            int safeSize = (userSize > MaxPageSize) ? MaxPageSize : userSize;
            
            for (int i = 0; i < safeSize; i++) // FALSE POSITIVE - Bounded
            {
                Response.Write("Row " + i + "<br/>");
            }
        }
        
        /// <summary>
        /// FALSE POSITIVE: If-else with const
        /// </summary>
        protected void ProcessIfElseBounded()
        {
            string retriesParam = Request.QueryString["retries"];
            int userRetries = int.Parse(retriesParam);
            
            int safeRetries;
            if (userRetries > MaxRetries)
                safeRetries = MaxRetries;
            else if (userRetries < 0)
                safeRetries = 0;
            else
                safeRetries = userRetries;
            
            for (int i = 0; i < safeRetries; i++) // FALSE POSITIVE - Bounded
            {
                Response.Write("Retry " + i + "<br/>");
            }
        }
        
        /// <summary>
        /// FALSE POSITIVE: Math.Min with long
        /// </summary>
        protected void ProcessLongBounded()
        {
            string iterParam = Request.QueryString["iterations"];
            long userIterations = long.Parse(iterParam);
            
            long safeIterations = Math.Min(userIterations, MaxIterations);
            
            for (long i = 0; i < safeIterations; i++) // FALSE POSITIVE - Bounded
            {
                Response.Write("Iteration " + i + "<br/>");
            }
        }
        
        /// <summary>
        /// FALSE POSITIVE: Both min and max bounds
        /// </summary>
        protected void ProcessDoubleBounded()
        {
            string batchParam = Request.QueryString["batchSize"];
            int userBatch = int.Parse(batchParam);
            
            int safeBatch = Math.Max(1, Math.Min(userBatch, MaxBatchSize));
            
            for (int i = 0; i < safeBatch; i++) // FALSE POSITIVE - Bounded
            {
                Response.Write("Batch " + i + "<br/>");
            }
        }
        
        /// <summary>
        /// FALSE POSITIVE: Clamp method with constants
        /// </summary>
        protected void ProcessClampBounded()
        {
            string valueParam = Request.QueryString["value"];
            int userValue = int.Parse(valueParam);
            
            int safeValue = Clamp(userValue, 0, MaxItems);
            
            for (int i = 0; i < safeValue; i++) // FALSE POSITIVE - Clamped
            {
                Response.Write("Value " + i + "<br/>");
            }
        }
        
        /// <summary>
        /// FALSE POSITIVE: Enum length with constant max
        /// </summary>
        protected void ProcessEnumBounded()
        {
            string levelParam = Request.QueryString["level"];
            int userLevel = int.Parse(levelParam);
            
            int enumLength = Enum.GetValues(typeof(Level)).Length;
            int safeLevel = Math.Min(userLevel, Math.Min(enumLength, MaxRetries));
            
            for (int i = 0; i < safeLevel; i++) // FALSE POSITIVE - Bounded
            {
                Response.Write("Level " + i + "<br/>");
            }
        }
        
        private static int Clamp(int value, int min, int max)
        {
            return Math.Max(min, Math.Min(value, max));
        }
        
        private enum Level
        {
            Low,
            Medium,
            High,
            Critical
        }
    }
}

File: src/main/java/com/checkmarx/falsepositive/loop/LoopCondition_FP_ImmutableConstants.java
package com.checkmarx.falsepositive.loop;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.io.PrintWriter;

/**
 * FALSE POSITIVE SCENARIOS: Unchecked Input for Loop Condition - Immutable Constants
 * 
 * Pattern: Loop bounds are constrained by final constants, ensuring safe iteration counts
 * 
 * Even though user input flows to the loop condition, it's bounded by immutable constants.
 * CxQL doesn't recognize Math.min/max with constants as safe bounds.
 * 
 * All scenarios are FALSE POSITIVES - SAFE bounded loops.
 */
public class LoopCondition_FP_ImmutableConstants {
    
    // Immutable constant limits
    private static final int MAX_ITEMS = 100;
    private static final int MAX_PAGE_SIZE = 50;
    private static final int MAX_RETRIES = 10;
    private static final int MAX_BATCH_SIZE = 1000;
    private static final long MAX_ITERATIONS = 500L;
    
    /**
     * FALSE POSITIVE: Math.min() with final constant
     * User input bounded by MAX_ITEMS constant (100)
     */
    public void processMathMinBounded(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String countParam = request.getParameter("count");
        int userCount = Integer.parseInt(countParam);
        
        // SAFE: Math.min ensures max of 100 iterations
        int safeCount = Math.min(userCount, MAX_ITEMS);
        
        PrintWriter out = response.getWriter();
        for (int i = 0; i < safeCount; i++) { // FALSE POSITIVE - Bounded by constant
            out.write("Item " + i + "\n");
        }
    }
    
    /**
     * FALSE POSITIVE: Ternary with final constant
     * User input bounded by ternary check against MAX_PAGE_SIZE
     */
    public void processTernaryBounded(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String sizeParam = request.getParameter("pageSize");
        int userSize = Integer.parseInt(sizeParam);
        
        // SAFE: Ternary ensures max of 50 iterations
        int safeSize = (userSize > MAX_PAGE_SIZE) ? MAX_PAGE_SIZE : userSize;
        
        PrintWriter out = response.getWriter();
        for (int i = 0; i < safeSize; i++) { // FALSE POSITIVE - Bounded by constant
            out.write("Row " + i + "\n");
        }
    }
    
    /**
     * FALSE POSITIVE: If-else with final constant
     * User input bounded by explicit if check
     */
    public void processIfElseBounded(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String retriesParam = request.getParameter("retries");
        int userRetries = Integer.parseInt(retriesParam);
        
        // SAFE: If check ensures max of 10 retries
        int safeRetries;
        if (userRetries > MAX_RETRIES) {
            safeRetries = MAX_RETRIES;
        } else if (userRetries < 0) {
            safeRetries = 0;
        } else {
            safeRetries = userRetries;
        }
        
        PrintWriter out = response.getWriter();
        for (int i = 0; i < safeRetries; i++) { // FALSE POSITIVE - Bounded by constant
            out.write("Retry " + i + "\n");
        }
    }
    
    /**
     * FALSE POSITIVE: Math.min with long constant
     * Long values bounded by MAX_ITERATIONS
     */
    public void processLongBounded(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String iterParam = request.getParameter("iterations");
        long userIterations = Long.parseLong(iterParam);
        
        // SAFE: Math.min ensures max of 500 iterations
        long safeIterations = Math.min(userIterations, MAX_ITERATIONS);
        
        PrintWriter out = response.getWriter();
        for (long i = 0; i < safeIterations; i++) { // FALSE POSITIVE - Bounded by constant
            out.write("Iteration " + i + "\n");
        }
    }
    
    /**
     * FALSE POSITIVE: Multiple constant bounds
     * User input bounded by both min and max constants
     */
    public void processDoubleBounded(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String batchParam = request.getParameter("batchSize");
        int userBatch = Integer.parseInt(batchParam);
        
        // SAFE: Both min and max bounds applied
        int safeBatch = Math.max(1, Math.min(userBatch, MAX_BATCH_SIZE));
        
        PrintWriter out = response.getWriter();
        for (int i = 0; i < safeBatch; i++) { // FALSE POSITIVE - Bounded by constants
            out.write("Batch item " + i + "\n");
        }
    }
    
    /**
     * FALSE POSITIVE: Clamp utility method
     * User input bounded by custom clamp method using constants
     */
    public void processClampBounded(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String valueParam = request.getParameter("value");
        int userValue = Integer.parseInt(valueParam);
        
        // SAFE: Clamp ensures value is within [0, MAX_ITEMS]
        int safeValue = clamp(userValue, 0, MAX_ITEMS);
        
        PrintWriter out = response.getWriter();
        for (int i = 0; i < safeValue; i++) { // FALSE POSITIVE - Clamped by constants
            out.write("Value " + i + "\n");
        }
    }
    
    /**
     * FALSE POSITIVE: Enum ordinal with constant max
     * Enum ordinal already bounded, plus constant check
     */
    public void processEnumOrdinalBounded(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String levelParam = request.getParameter("level");
        int userLevel = Integer.parseInt(levelParam);
        
        // SAFE: Bounded by enum size and constant
        int safeLevel = Math.min(userLevel, Math.min(Level.values().length, MAX_RETRIES));
        
        PrintWriter out = response.getWriter();
        for (int i = 0; i < safeLevel; i++) { // FALSE POSITIVE - Bounded by enum + constant
            out.write("Level " + i + "\n");
        }
    }
    
    // Helper method
    private static int clamp(int value, int min, int max) {
        return Math.max(min, Math.min(value, max));
    }
    
    private enum Level {
        LOW, MEDIUM, HIGH, CRITICAL
    }
}

